leshan-core/src/main/java/org/eclipse/leshan/core/response/ObserveResponse.java

```diff
@@ -40,17 +40,34 @@ public ObserveResponse(ResponseCode code, LwM2mNode content, List<TimestampedLwM
 
     public ObserveResponse(ResponseCode code, LwM2mNode content, List<TimestampedLwM2mNode> timestampedValues,
             SingleObservation observation, String errorMessage, Object coapResponse) {
-        super(code, timestampedValues != null && !timestampedValues.isEmpty() ? timestampedValues.get(0).getNode()
-                : content, errorMessage, coapResponse);
+        super(code, getContent(content, timestampedValues), errorMessage, coapResponse);
 
         // CHANGED is out of spec but is supported for backward compatibility. (previous draft version)
         if (ResponseCode.CHANGED.equals(code)) {
-            if (content == null)
+            if (this.content == null)
                 throw new InvalidResponseException("Content is mandatory for successful response");
         }
 
         this.observation = observation;
-        this.timestampedValues = timestampedValues;
+        this.timestampedValues = hasSingleItemWithoutTimestamp(timestampedValues) ? null : timestampedValues;
+    }
+
+    private static LwM2mNode getContent(LwM2mNode content, List<TimestampedLwM2mNode> timestampedValues) {
+        if (hasSingleItemWithoutTimestamp(timestampedValues)) {
+            content = timestampedValues.get(0).getNode();
+            timestampedValues = null;
+        }
+
+        return getFirstListItemOrContent(content, timestampedValues);
+    }
+
+    private static LwM2mNode getFirstListItemOrContent(LwM2mNode content,
+            List<TimestampedLwM2mNode> timestampedValues) {
+        return timestampedValues != null && timestampedValues.size() > 0 ? timestampedValues.get(0).getNode() : content;
+    }
+
+    private static boolean hasSingleItemWithoutTimestamp(List<TimestampedLwM2mNode> timestampedValues) {
+        return timestampedValues != null && timestampedValues.size() == 1 && !timestampedValues.get(0).isTimestamped();
     }
```

Comment:
Does it change the behavior or is it just because you think the code is cleaner like this ?

Reply:
I moved the logic from ObservationServiceImpl.createObserveResponse():
```java
// create lwm2m response
if (timestampedNodes.size() == 1 && !timestampedNodes.get(0).isTimestamped()) {
    return new ObserveResponse(toLwM2mResponseCode(coapResponse.getCode()),
            timestampedNodes.get(0).getNode(), null, observation, null, coapResponse);
} else {
    return new ObserveResponse(toLwM2mResponseCode(coapResponse.getCode()), null, timestampedNodes,
            observation, null, coapResponse);
}
```
into ObserveResponse class.
I don't have strong opinion about that if the code is cleaner now. It's probably less cleaner than I intended :)

Reply:
I'm not sure too. :sweat_smile: 

Generally if benefits is not so clear I prefer to not change the code so much.
(If there is clear benefits I think its important to change the code to make it cleaner)

For this one, I let you decide :)